lib/spack/spack/fetch_strategy.py

```diff
@@ -1514,8 +1514,15 @@ def for_package_version(pkg, version):
 
     check_pkg_attributes(pkg)
 
-    if not isinstance(version, spack.version.VersionBase):
-        version = spack.version.Version(version)
+    if version is not None:
```

Comment:
I don't follow this, can you at least add a comment as to why this is mutating `pkg.spec`?

Reply:
The issue here is the following. When we initialize a `Spec` just by name we do:

https://github.com/spack/spack/blob/18e0b893f268443d4d285bcfa7c071f4838c5557/lib/spack/spack/spec.py#L1265

There are contexts in which we want to extrapolate a version `url` from an abstract spec with just the name. Sometimes this means calling `url_for_version`, which has an unfortunate API:

https://github.com/spack/spack/blob/18e0b893f268443d4d285bcfa7c071f4838c5557/var/spack/repos/builtin/packages/cddlib/package.py#L24-L29

where we pass the version as a string. Thus for any comparison we relied on `satisfies(..., strict=False)` i.e. the current `intersects`. 

I think we don't want to have changes in packages to fit an implementation of extrapolation that can be improved later, and `satisfies` gives the same result as `intersects` for concrete specs. Thus, in contexts where we call this function with an abstract spec I currently hack the version to make it satisfy the correct branch in `url_for_version`.

Reply:
Added a comment and an assert (for safety) in https://github.com/spack/spack/pull/35681/commits/a1c1ecef7851343413af7810b8326a3e5ef95d55